src/lib/recurringPurposes.js

```diff
@@ -275,27 +386,99 @@ const getTimeserieWithPurpose = timeseries => {
   })
 }
 
-const isLoopTrip = timeserie => {
-  return (
-    timeserie?.aggregation?.startPlaceDisplayName ===
-    timeserie?.aggregation?.endPlaceDisplayName
-  )
-}
-
 export const findPurposeFromSimilarTimeserieAndWaybacks = async (
   client,
   timeserie
 ) => {
-  const similarTimeseries = await findSimilarTimeseries(client, timeserie)
+  const similarTimeseries = await findSimilarRecurringTimeseries(
+    client,
+    timeserie
+  )
+  log('info', `Found ${similarTimeseries.length} similar timeseries`)
+
   let tsWithPurpose = getTimeserieWithPurpose(similarTimeseries)
+
   if (!tsWithPurpose) {
     // if no similar trips with purpose found, try with waybacks trips
-    const waybackTimeseries = await findWaybackTimeseries(client, timeserie)
+    const waybackTimeseries = await findSimilarRecurringTimeseries(
+      client,
+      timeserie,
+      {
+        isWayBack: true
+      }
+    )
+    log('info', `Found ${waybackTimeseries.length} wayback timeseries`)
     tsWithPurpose = getTimeserieWithPurpose(waybackTimeseries)
   }
   return tsWithPurpose?.aggregation?.purpose || null
 }
 
+/**
+ * Try to set a purpose to a timeserie, from existing contacts' addresses
+ *
+ * It is based on the geo coordinates. If a match is found on the start or end
+ * address, a relationship is added, and the contact geo information is updated
+ * with the trip coordinates, to incrementally improve the geolocation precision.
+ *
+ * @param {object} client - The cozy client instance
+ * @param {object} timeserie - The timeserie to categorize
+ * @param {Array<object>} contacts - The contacts with geo information
+ * @returns {object} The timeserie to update
+ */
+const findPurposeFromContactAddresses = async (client, timeserie, contacts) => {
+  const { matchingStart, matchingEnd } = await findStartAndEnd(
+    timeserie,
+    contacts
+  )
+  let newTimeserie = { ...timeserie }
+  if (!matchingStart && !matchingEnd) {
+    log('info', 'No matching start and end places for this trip')
+    return newTimeserie
```

Comment:
nit: no need to spread in that case

Reply:
I feel like this is safer to do a shallow copy, rather than a reference copy, don't you think? Reference copy can be tedious if we don't pay attention and update the object at several places